### ISSUE TITLE
How to support CoAP?

### ISSUE THREAD

**Issue by tkellogg** [jdegoes/blueeyes#133]
I'd like to explore the possibility of having a web framework support both HTTP and [CoAP](http://tools.ietf.org/search/draft-ietf-core-coap-18). How suitable would Blueeyes be to support an extension that handles both CoAP and HTTP requests and responses for the same resources? My idea would be to use [Californium](https://github.com/mkovatsc/Californium) as the CoAP implementation. Even though it's written in Java, Californium should be mostly friendly toward being wrapped in a functional style, except for it's pervasive use of mutable data.

CoAP is a RESTful protocol that mirrors a lot of the functionality of HTTP/1.1 but as a compact binary protocol suitable for embedded devices. It runs over UDP, but the Californium implementation takes care of ACKs and resending messages. The biggest incompatibility is the fact that there aren't any custom CoAP headers and many of the normal HTTP headers don't have a CoAP analogy.

I'd like to pursue this, but I want to know your thoughts. Would this be difficult? Is blueeyes a suitable choice for this?

**Comment by jdegoes** [comment]
RedEyes (AKA BlueEyes 2) would be a better fit for this, as it cleanly separates the specification of an API from its compilation to an executable format (this choice also allows but does not require the use of mutable implementation). The real question is, what would be the benefit? The ability to deploy both an HTTP API and a CoAP API simultaneously, from the same description?

**Comment by tkellogg** [comment]
Yes, precisely. CoAP endpoints are usually described in similar terms, with methods and URIs. There's a lot of overlap between how you interface with both protocols. Where is the repository for RedEyes?

**Comment by jdegoes** [comment]
It's not public yet, but it will be at github.com/jdegoes/redeyes/ within the month.
